Content.Server/Zombies/ZombieSystem.cs

```diff
@@ -74,18 +81,38 @@ public override void Update(float frameTime)
                 // 1x at 30s, 3x at 60s, 6x at 90s, 10x at 120s.
                 var pain_multiple = 0.1 + 0.02 * comp.InfectedSecs + 0.0005 * comp.InfectedSecs * comp.InfectedSecs;
                 comp.NextTick = curTime;
-                _damageable.TryChangeDamage(uid, comp.Damage * pain_multiple, true, false);
+                _damageable.TryChangeDamage(uid, comp.Damage * pain_multiple, true, false, damage);
             }
 
-            var zomb_query = EntityQueryEnumerator<ZombieComponent>();
             // Heal the zombified
-            while (zomb_query.MoveNext(out var uid, out var comp))
+            while (zombQuery.MoveNext(out var uid, out var comp, out var damage, out var mobState))
             {
+                // Process only once per second
                 if (comp.NextTick + TimeSpan.FromSeconds(1) > curTime)
                     continue;
 
                 comp.NextTick = curTime;
-                _damageable.TryChangeDamage(uid, comp.Damage, true, false);
+
+                if (comp.Permadeath)
+                {
+                    // No healing
+                    continue;
+                }
+
+                // Healing increases over time for zombies currently in crit
+                if (mobState.CurrentState == MobState.Alive)
+                {
+                    comp.SecondsCrit = 0;
+                }
+                else
+                {
+                    comp.SecondsCrit += 1;
+                }
+
+                // Healing increases over 50 seconds to a maximum of 10x rate. They will be half healed by then at least.
+                //   At that rate most zombies will revive from the remaining 50 (of 100) damage after a further 25sec
+                float healMultiple = 1.0f + Math.Min(comp.SecondsCrit * 0.2f, 20.0f);
+                _damageable.TryChangeDamage(uid, comp.Damage * healMultiple, true, false, damage);
```

Comment:
reset SecondsCrit on the MobStateChanged handler
```suggestion
                if (mobState.CurrentState != MobState.Critical)
                	continue;

                comp.SecondsCrit += 1;
                // Healing increases over 50 seconds to a maximum of 10x rate. They will be half healed by then at least.
                //   At that rate most zombies will revive from the remaining 50 (of 100) damage after a further 25sec
                float healMultiple = 1.0f + Math.Min(comp.SecondsCrit * 0.2f, 20.0f);
                _damageable.TryChangeDamage(uid, comp.Damage * healMultiple, true, false, damage);
```

Reply:
I tried that, but it just seemed like hidden behaviour to spread the code that changed that parameter around. Right now only this method is writing or reading from that param and I like that.